httpclient5-testing/src/main/java/org/apache/hc/client5/testing/async/ServiceUnavailableAsyncDecorator.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+
+package org.apache.hc.client5.testing.async;
+
+import org.apache.hc.core5.function.Resolver;
+import org.apache.hc.core5.http.EntityDetails;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.HttpException;
+import org.apache.hc.core5.http.HttpHeaders;
+import org.apache.hc.core5.http.HttpRequest;
+import org.apache.hc.core5.http.HttpResponse;
+import org.apache.hc.core5.http.HttpStatus;
+import org.apache.hc.core5.http.HttpVersion;
+import org.apache.hc.core5.http.ProtocolVersion;
+import org.apache.hc.core5.http.message.BasicHttpResponse;
+import org.apache.hc.core5.http.nio.AsyncServerExchangeHandler;
+import org.apache.hc.core5.http.nio.CapacityChannel;
+import org.apache.hc.core5.http.nio.DataStreamChannel;
+import org.apache.hc.core5.http.nio.ResponseChannel;
+import org.apache.hc.core5.http.protocol.HttpContext;
+import org.apache.hc.core5.util.Args;
+import org.apache.hc.core5.util.TimeValue;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class ServiceUnavailableAsyncDecorator implements AsyncServerExchangeHandler {
+
+    private final AsyncServerExchangeHandler exchangeHandler;
+    private final Resolver<HttpRequest, TimeValue> serviceAvailabilityResolver;
+    private final AtomicBoolean serviceUnavailable;
+
+    public ServiceUnavailableAsyncDecorator(final AsyncServerExchangeHandler exchangeHandler,
+                                            final Resolver<HttpRequest, TimeValue> serviceAvailabilityResolver) {
+        this.exchangeHandler = Args.notNull(exchangeHandler, "Exchange handler");
+        this.serviceAvailabilityResolver = Args.notNull(serviceAvailabilityResolver, "Service availability resolver");
+        this.serviceUnavailable = new AtomicBoolean();
+    }
+
+    @Override
+    public void handleRequest(final HttpRequest request,
+                              final EntityDetails entityDetails,
+                              final ResponseChannel responseChannel,
+                              final HttpContext context) throws HttpException, IOException {
+        final TimeValue retryAfter = serviceAvailabilityResolver.resolve(request);
+        serviceUnavailable.set(TimeValue.isPositive(retryAfter));
+        if (serviceUnavailable.get()) {
+            final HttpResponse response = new BasicHttpResponse(HttpStatus.SC_SERVICE_UNAVAILABLE);
+            response.addHeader(HttpHeaders.RETRY_AFTER, Long.toString(retryAfter.toSeconds()));
+            final ProtocolVersion version = request.getVersion();
+            if (version != null && version.compareToVersion(HttpVersion.HTTP_2) < 0) {
```

Comment:
Why close if version is 1.1+?

Reply:
@michael-o It is the other way around. Close connection If the connection is HTTP/1.1 or lower. HTTP/2 connections do not support `Connection` directives.

Reply:
Sorry, my question was misleading. Once again: Why close the connection of th version is exactly 1.1?

Reply:
@michael-o What is the point of keeping the connection alive if the server is unable to serve requests over it?

Reply:
Well, from an abstract point of view this connection is used to a host:port pair, but the URI path can be proxied to to other backend servers as well. E.g., `example.com/service1` and `example.com/service2` can be to completely different backends where service1 is offline, but service2 not.

Reply:
@michael-o I am not sure I understand what it is exactly you are proposing or objecting to.

Reply:
I am neither, I am just trying to clarify that the status code from the server applies solely to the path of the request and it does not give you any idicator that anything else served by this server is unavailable thus making the entire connection invalid . I don't consider the connection to be invalid unless TCP tells me otherwise or the server tells me to close.

Reply:
@michael-o This is a test class. It is there to merely validate the assumptions of the test case.

Reply:
Duh, I am an idiot. I didn't notice this. So this spot is fine. Sorry for the noise.

Reply:
@michael-o Not at all. Thank you for a thorough review!